packages/cli/src/util/update-notifier.ts

```diff
@@ -0,0 +1,107 @@
+import type { PackageJson } from '@vercel/build-utils';
+import { mkdirpSync, readJSONSync, writeJSONSync } from 'fs-extra';
+import { Agent as HttpsAgent } from 'https';
+import fetch from 'node-fetch';
+import { join } from 'path';
+import XDGAppPaths from 'xdg-app-paths';
+
+interface UpdateNotifierConfig {
+  version: string;
+  expireAt: number;
+  notified: boolean;
+}
+
+const xdgDir = XDGAppPaths('com.vercel.cli').cache();
+const cacheDir = join(xdgDir, 'update-notifier');
+
+export class UpdateNotifier {
+  private pkg: PackageJson;
+  private distTag: string;
+  private updateCheckInterval: number;
+  private cacheFile: string;
+
+  constructor({
+    pkg,
+    distTag,
+    updateCheckInterval,
+  }: {
+    pkg: PackageJson;
+    distTag: string;
+    updateCheckInterval: number;
+  }) {
+    this.pkg = pkg;
+    this.distTag = distTag;
+    this.updateCheckInterval = updateCheckInterval;
+    this.cacheFile = join(cacheDir, `${pkg.name}-${distTag}.json`);
+  }
+
+  getLastestVersion(): string | null {
+    const { pkg, cacheFile } = this;
+
+    let config: UpdateNotifierConfig;
+    try {
+      config = readJSONSync(cacheFile);
+    } catch (e) {
+      console.error(`failed to read ${cacheFile}`);
+      return null;
+    }
+
+    if (
+      config.version !== pkg.version &&
+      !config.notified &&
+      process.stdout.isTTY
```

Comment:
If we want to write unit tests for this, we probably want `client.stdout.isTTY`.

Reply:
The `isTTY` check has been moved to the CLI's `index.ts` as to decouple the update notifier from the console.